Modules/_functoolsmodule.c

```diff
@@ -1281,37 +1280,45 @@ lru_cache_prepend_link(lru_cache_object *self, lru_list_elem *link)
    so that we know the cache is a consistent state.
  */
 
-static PyObject *
-bounded_lru_cache_wrapper(lru_cache_object *self, PyObject *args, PyObject *kwds)
+static int
+bounded_lru_cache_wrapper_pre_call_lock_held(lru_cache_object *self, PyObject *args, PyObject *kwds,
+                                             PyObject **result, PyObject **key, Py_hash_t *hash)
 {
     lru_list_elem *link;
-    PyObject *key, *result, *testresult;
-    Py_hash_t hash;
 
-    key = lru_cache_make_key(self->kwd_mark, args, kwds, self->typed);
-    if (!key)
-        return NULL;
-    hash = PyObject_Hash(key);
-    if (hash == -1) {
-        Py_DECREF(key);
-        return NULL;
+    PyObject *key_ = *key = lru_cache_make_key(self->kwd_mark, args, kwds, self->typed);
+    if (!key_)
+        return -1;
+    Py_hash_t hash_ = *hash = PyObject_Hash(key_);
+    if (hash_ == -1) {
+        Py_DECREF(key_);
+        return -1;
     }
-    link  = (lru_list_elem *)_PyDict_GetItem_KnownHash(self->cache, key, hash);
+    link = (lru_list_elem *)_PyDict_GetItem_KnownHash(self->cache, key_, hash_);
     if (link != NULL) {
         lru_cache_extract_link(link);
         lru_cache_append_link(self, link);
-        result = link->result;
+        *result = link->result;
         self->hits++;
-        Py_INCREF(result);
-        Py_DECREF(key);
-        return result;
+        Py_INCREF(link->result);
+        Py_DECREF(key_);
+        return 1;
     }
     if (PyErr_Occurred()) {
-        Py_DECREF(key);
-        return NULL;
+        Py_DECREF(key_);
+        return -1;
     }
     self->misses++;
-    result = PyObject_Call(self->func, args, kwds);
+    return 0;
+}
+
+static PyObject *
+bounded_lru_cache_wrapper_post_call_lock_held(lru_cache_object *self,
```

Comment:
Maybe use more meaningful names `bounded_lru_cache_get` and `bounded_lru_cache_update`?